utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,601 @@
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map (makeVertexInputBindingDescription . calculate)
+    . groupBy ((==) `on` fst)
+    . sortOn fst
+    . map extract
+    . V.toList
+ where
+  extract :: VertexAttribute -> ("binding" ::: Int, "size" ::: Int)
+  extract = liftA2 (,)
+                   (fromIntegral . (binding :: VertexAttribute -> Word32))
+                   (fromIntegral . size)
+  calculate
+    :: [("binding" ::: Int, "size" ::: Int)]
+    -> ("binding" ::: Int, "stride" ::: Int)
+  calculate = liftA2 (,) (fst . head) (sum . (snd <$>))
+
+makeVertexInputBindingDescription
+  :: ("binding" ::: Int, "stride" ::: Int) -> VertexInputBindingDescription
+makeVertexInputBindingDescription (binding, stride) = zero
+  { binding   = fromIntegral binding
+  , stride    = fromIntegral stride
+  , inputRate = VERTEX_INPUT_RATE_VERTEX
+  }
+
+makeVertexAttribute :: Input -> Vector VertexAttribute
+makeVertexAttribute Input {..} = V.generate
+  count
+  (\i -> VertexAttribute { binding  = 0
+                         , location = fromIntegral . (+ location) $ i
+                         , size
+                         , format
+                         }
+  )
+ where
+  count          = maybe 1 V.sum array :: Int
+  (size, format) = convertVertexAttributeType type'
+
+makeVertexInputAttributeDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputAttributeDescription
+makeVertexInputAttributeDescriptions =
+  V.fromList
+    . join
+    . map process
+    . groupBy ((==) `on` (binding :: VertexAttribute -> Word32))
+    . V.toList
+ where
+  process :: [VertexAttribute] -> [VertexInputAttributeDescription]
+  process = snd . foldl'
+    (\(nextOffset, acc) cur -> liftA2
+      (,)
+      fst
+      ((acc ++) . pure . snd)
+      (makeVertexInputAttributeDescription nextOffset cur)
+    )
+    (0, [])
+
+makeVertexInputAttributeDescription
+  :: Word32 -> VertexAttribute -> (Word32, VertexInputAttributeDescription)
+makeVertexInputAttributeDescription offset VertexAttribute {..} =
+  (offset + size, zero { binding, location, offset, format })
+
+readProcessHandler
+  :: MonadIO m
+  => (ExitCode, BL.ByteString, BL.ByteString)
+  -> ExceptT String m BL.ByteString
+readProcessHandler (exitCode, result, err) = case exitCode of
+  ExitFailure errCode ->
+    throwE $ "errCode: " <> show errCode <> ", " <> BL.unpack
+      (if BL.null err then result else err)
+  ExitSuccess -> pure result
+
+reflect
+  :: MonadIO m
+  => ShaderStage
+  -> "code" ::: Text
+  -> m (Either String (B.ByteString, BL.ByteString))
+reflect shaderStage code =
+  liftIO . runExceptT . withSystemTempDirectory "th-spirv" $ \dir -> do
+    let stage  = T.unpack . shaderStageToText $ shaderStage
+    let shader = dir </> stage <.> "glsl"
+    let spv    = dir </> stage <.> "spv"
+    liftIO . T.writeFile shader $ code
+    _spv <-
+      readProcessHandler
```

Comment:
`readProcessHandler` can take the command and its arguments and run the process itself, so you won't have to duplicate this pipeline 3 times.

Reply:
a function `procE :: MonadIO m => FilePath -> [String] -> ExceptT String m BL.ByteString` has been add to reduce the duplicate code